### ISSUE TITLE
Render Planet

### ISSUE THREAD

**Issue by lukasmartinelli** [osm2vectortiles/osm2vectortiles#251]
Currently running Europe (based on latest master so v1.4 + patches) to detect problems in the export process before we start the initial world rendering once the stable v1.5 is released.

Will go through the entire process (including updating once, dirty tiles and again workers) with Europe as test.

Already detected some bugs in 195bbfa9f21899e113e6c011a085ba3d9f03e91b and c67170bdf603291db7564325e744e62b7f3bdede.

The nice thing is that **tilelive-copy** seems to pump out much more tiles per second than **tl** before did. (80/s for one machine).
Other interesting fact is that once you scale to two workers the tiles per second for the workers just half - which means that they already pretty much use the database at full limit.

**Comment by lukasmartinelli** [comment]
Had a bit problem with RabbitMQ. It actually has good support for long running tasks but requires something called *heartbeats* to continously send *heartbeats* during a long running task to tell that you are still ok (and especially to keep the TCP connections open which can sometimes be closed by network devices).

I now disabled heartbeats since it works just fine without them (tasks max go 10minutes) - RabbitMQ detects when clients crash when they close the connection channel.
Had to fork the rabbitmq Docker image to be able to configure heartbeats https://github.com/osm2vectortiles/rabbitmq.

Was a bit trouble but still think it is totally worth it out of the single reason that anyone is theoretically able to run the clustered setup without relying on a AWS account.

**Comment by klokan** [comment]
We have used [beanstalkd](http://kr.github.io/beanstalkd/) for similar job management on large raster processing tasks.

It has exactly what a similar project with long running (hours) tasks pulled from a job queue needs - and is designed to run exactly this sort of task well. No long TCP connections open for acknowledgment, no heartbeats, no extra functionality and optimization for different use-cases such as general message queue, possibility to ask for a status of a particular task, etc. 

Here is a picture of the typical job lifecycle:

````
   put            reserve               delete
  -----> [READY] ---------> [RESERVED] --------> *poof*
```


Here is a picture with more possibilities:
```
   put with delay               release with delay
  ----------------> [DELAYED] <------------.
                        |                   |
                        | (time passes)     |
                        |                   |
   put                  v     reserve       |       delete
  -----------------> [READY] ---------> [RESERVED] --------> *poof*
                       ^  ^                |  |
                       |   \  release      |  |
                       |    `-------------'   |
                       |                      |
                       | kick                 |
                       |                      |
                       |       bury           |
                    [BURIED] <---------------'
                       |
                       |  delete
                        `--------> *poof*

```

You could also prioritize certain tasks (such give higher priority to rendering diff tiles, if another rendering runs - so it is not strictly FIFO).

See for a more complete overview: https://www.digitalocean.com/community/tutorials/how-to-install-and-use-beanstalkd-work-queue-on-a-vps

There are several client libraries and graphical web user interfaces:
https://github.com/kr/beanstalkd/wiki/client-libraries

**Comment by lukasmartinelli** [comment]
Yes any queue qualifies that supports:

- Persistent messages
- Long running tasks
- Some sort of ACK mechanism to prevent message loss
- Plays nice with Docker

And at osm2vectortiles we don't need a crazy enterprise message queue. I just want something simple decoupled from AWS - beanstalkd might fit perfect there.

You told about a wrapper around different message queue backends you use internally?
:+1: if you ever get around releasing it as OSS.

Pioritizing would in our case be kind of a cool feature - we don't require FIFO.
From making osm2vectortiles updateable we also gain the capability to fix local errors like #87 easily by just scheduling areas as new jobs.

**Comment by klokan** [comment]
Here you have it:
https://github.com/klokantech/cloudwrapper/blob/master/btd.py

The interface follows the official python Queue class.
Please use it directly or as a fork - as a submodule in OSM2VectorTiles codebase.
Pull request and improvements are very welcome.

Next to the native queues for Amazon / Google it does contain also logging and metrics.
For open-source / private cloud the queues mean Beanstalkd and metrics & logging mean InfluxDB (WIP).

**Comment by lukasmartinelli** [comment]
Took 46h for rendering the 1024 pyramid tasks for the 4/2/3 tile for a single machine.
Don't remember whether that is fast or slow compared to previously.

<img width="603" alt="tiles_a_la_google_maps__coordinates__tile_bounds_and_projection_-_conversion_to_epsg_900913__epsg_3785__and_epsg_4326__wgs84_" src="https://cloud.githubusercontent.com/assets/1288339/14681926/7a69d7c8-0723-11e6-843f-1f4053b73d32.png">

**Comment by lukasmartinelli** [comment]
Doing another import and export to see whether it is faster now especially because of #261 and #257.
Some pyramid tasks that took hours should now be much faster.

I created a tag [v1.4.1](https://github.com/osm2vectortiles/osm2vectortiles/releases/tag/v1.4.1) just for rendering so I know what version I am rendering with.
I hope this doesn't screw up your cool GitHub release script - if it does I can delete the tag later on.

**Comment by lukasmartinelli** [comment]
Doing another export since we met the other performance problems #275 and since the mapping has changed again. The place labels in #275 where making the export really really slow.

Want to switch to beanstalkd some time in the future and working on https://github.com/lukasmartinelli/cloudwrapper to create a pip installable Python module that I can use in `export_remote`. But at the moment other prios and RabbitMQ works well enough for testing now.

**Comment by lukasmartinelli** [comment]
Just want to note it down somewhere. World import with imposm3 **in diff mode!** with v1.4.2 takes 14hrs including creating the GIST indizes and clustering on geohash.

And then we have our extra scripts that create some tables and additional indizes, update the timestamp that take up 1-2 additional hours.

**Comment by lukasmartinelli** [comment]
Just for future reference.

The **limiting factor on worker machines is actually CPU**. We long believed that we need high-memory machines (like you expect for PostgreSQL to keep all indizes in memory). But it turns out that the limiting factor for PostgreSQL in our use case is CPU. On the HSR dev server with 40 cores PostgreSQL can serve many more workers (and using the same memory) by using 12 cores to serve 4 workers (each one using one core fully). 

This is good to know because until know we use 50GB memory and 8 core servers as workers we might better use 50GB memory and 16 core servers.

**Comment by lukasmartinelli** [comment]
Using the HSR server we can approximately process 100 x z8 pyramids per hour (expect another speedup after #306. This means using just this server we can be done with the 65k pyramid jobs in 650 hrs (approx. 4 weeks). To get this down to a week we probably need 4 x (12 core, 50GB) machines which is doable. This also means that there is a chance that a single server can keep up with the updates (which would be a huge win and saves money).

**Comment by klokan** [comment]
Sounds great. Cool!

**Comment by lukasmartinelli** [comment]
Will continue in this issue rendering the planet (or at least starting now and check how it performs with several machines). Apart from #304 **avoid changing the mapping now**. We can still do another import if absolutely necessary but the time to freeze it has come now.

Steps now.

- [x] Download world PBF `ab520bbadd3e80c07cd7bc2b761b654a` from `2016-04-27` to shared disk
- [x] Mount shared disk on all workers
- [ ] *in progress*: Import world PBF on each worker using the latest mapping from master  and using the newest id mapping from #304 - this will take a day.

Previously I imported only on one machine and then shared the entire database. But just importing on each machine is a bit more practical.

- [ ] Start up the RabbitMQ server on the master, open network ports (also that the management portal of RabbitMQ can be opened for tracking progress).
- [ ] Schedule jobs for the quarter planet at first
- [ ] Ensure workers are able to connect to RabbitMQ and S3
- [ ] Run export-worker and scale them so that all cores on the workers are maxed out
- [ ] On the master run the merge-jobs component to continously merge the completed jobs back into one large PBF
- [ ] Monitor progress on the RabbitMQ mgmt portal

After this works nicely I will then remove the jobs /results/result PBF again, apply the latest changes from the PRs #311 #306 #304 and issues #307 start again with the real thing.

We should then runn import-diffs already the first time to keep up with the changes since 2016-04-27.

**Comment by lukasmartinelli** [comment]
Started rendering - let's see how fast 4 workers crunch through the 1024 tiles example I used for europe.

Perhaps we can even switch to high CPU machines instead of normal ones. The 16 cores are maxed out but memory not (the first machine CPU is only 50% on PostGIS but that was just bad timing/anomaly).

<img width="1274" alt="1__me_tiler-master___osm2vectortiles__ssh_" src="https://cloud.githubusercontent.com/assets/1288339/15020705/89af329a-1223-11e6-90ff-0ebe59692762.png">

**Comment by lukasmartinelli** [comment]
> Using the HSR server we can approximately process 100 x z8 pyramids per hour (expect another speedup after #306. This means using just this server we can be done with the 65k pyramid jobs in 650 hrs (approx. 4 weeks).

This still holds true for the Klokantech infra, actually better now with 120 z8 pyramids per hour so 480 z8 pyramids per hour for the 4 workers. 
A z8 pyramid has `4^0+4^1+4^2+4^3+4^5+4^6+4^7= 21589 ` tiles.

Notes for myself for what I already scheduled, will continue running it during the holidays so we can look at the results next week.

<img width="944" alt="tiles_a_la_google_maps__coordinates__tile_bounds_and_projection_-_conversion_to_epsg_900913__epsg_3785__and_epsg_4326__wgs84_" src="https://cloud.githubusercontent.com/assets/1288339/15024982/6cdde4e0-1237-11e6-9b48-49cfc26486bb.png">

Found very few z8 queries in road layer that are slow. They take 3.1s - but they are very rare and only for zoom level 8 (` z(2.18392e+06)=8`).
And the slowest part is actually the **Bitmap Index Scan** on the `road_class` index which I am not sure whether we can make this faster.

```sql
SELECT osm_ids2mbid(MAX(osm_id), false) AS osm_id, ST_CollectionExtract(ST_Collect(geometry), 2) AS geometry,
  road_type(road_class(type, NULL, NULL), type, NULL, NULL, NULL) AS type,
  road_class(type, NULL, NULL) AS class, road_oneway(0) AS oneway, 'none' AS structure, NULL AS z_order
  FROM (
    SELECT * FROM road_z5
    WHERE z(2.18392e+06) = 5
    UNION ALL
    SELECT * FROM road_z6toz7
    WHERE z(2.18392e+06) BETWEEN 6 AND 7
    UNION ALL
    SELECT * FROM road_z8toz9
    WHERE z(2.18392e+06) BETWEEN 8 AND 9
    UNION ALL
    SELECT * FROM road_z10
    WHERE z(2.18392e+06) = 10
  ) AS road_grouped_zoom_levels
  WHERE geometry && ST_SetSRID('BOX3D(3911129.863295898 4537301.999008063,4072564.867034191 4698737.002746355)'::box3d, 3857)
  GROUP BY type
```

<img width="1030" alt="dgf___explain_depesz_com" src="https://cloud.githubusercontent.com/assets/1288339/15025498/cb9eb5fc-1239-11e6-8f9c-363990eaec1f.png">

**Comment by sfkeller** [comment]
Seems to me impossible to speed up a Bitmap Index Scan.

The only thing which stroke my eyes is that you are generating a 3D(?) bbox like this:
`WHERE geometry && ST_SetSRID('BOX3D(3911129.863295898 4537301.999008063, 4072564.867034191 4698737.002746355)'::box3d, 3857)
`
What about this (box2d)?
`WHERE geometry && ST_SetSRID('BOX(3911129.863295898 4537301.999008063, 4072564.867034191 4698737.002746355)'::box2d, 3857)
`
... or even this (no string parsing and casting needed)?
`WHERE geometry && ST_MakeEnvelope(3911129.863295898, 4537301.999008063, 4072564.867034191, 4698737.002746355, 3857)
`

**Comment by lukasmartinelli** [comment]
> Seems to me impossible to speed up a Bitmap Index Scan.

Thanks for the feedback. On z8 the queries cover very big areas.

> The only thing which stroke my eyes is that you are generating a 3D(?) bbox like this

The `geometry && ST_SetSRID('BOX3D(3911129.863295898 4537301.999008063,4072564.867034191 4698737.002746355)'::box3d, 3857)`  part is actually inserted by Mapnik https://github.com/mapnik/mapnik/wiki/PostGIS.

We write our queries like this `geometry && !bbox!` and Mapnik will replace the bbox placehooder `!bbox!`, so this is out of our control. 
But this affects only query parsing speed.

**Comment by sfkeller** [comment]
Regarding Bitmap Index Scan just one question: It seems to me that the query you cited above does not correspond with the query shown in the EXPLAIN ANALYZE image you cite too: The Bitmap Index Scan in the EXPLAIN ANALYZE image shows "= ANY ('{motorway,moto...". What's about this ANY clause?

**Comment by lukasmartinelli** [comment]
> It seems to me that the query you cited above does not correspond with the query shown in the EXPLAIN ANALYZE image you cite too

Exactly. This is because there is also the view `road_z8toz9` involved which looks like this and is using the road_class index.
Previously this was horribly slow until we made an index on `road_class(type, service, access)`.

Now the query planner must decide how he is using the GIST index and road_class index to filter out the rows that are not relevant. In the case above he thinks road_class index is better applying before GIST (from what I guess).

```sql
CREATE OR REPLACE VIEW road_z8toz9 AS
    SELECT id AS osm_id, geometry, type, construction, tracktype, service, access, oneway, 'none'::varchar(4) AS structure, z_order
    FROM osm_road_geometry
    WHERE road_class(type, service, access) IN ('motorway', 'motorway_link', 'trunk', 'primary', 'secondary', 'major_rail');
```

**Comment by lukasmartinelli** [comment]
Okay dealing with the #314 does not have a straightforward solution. Since these are all edge cases I deal for now with them by putting timed out tilelive calls to a `failed-jobs` queue and deal with them later on. https://github.com/osm2vectortiles/osm2vectortiles/commit/4010e0c1921e6534d01830add8690d746ac40e8c

This way it least the workers are not stalled with the same failing message over and over again.

<img width="792" alt="rabbitmq_management" src="https://cloud.githubusercontent.com/assets/1288339/15073796/d44e3282-139b-11e6-95d4-eb7fdd7cf3d6.png">

**Comment by muesliq** [comment]
Hey guys, seems you are well in time for v1.6. Any idea when a new planet download will be available?

**Comment by manuelroth** [comment]
Thanks for asking, the new planet file with extracts and updated documentation should be available at the end of may.

**Comment by ImreSamu** [comment]
> Regarding Bitmap Index Scan ..

my optimization idea  ( need more research , and for  just the v2.x iterations  ) 

**IDEA:**   Data processing optimized imposm3 mappings ( = more `logical partitions` )
The current design is  `one database table per layer` sometimes not optimal,
but we can split the big  massive tables  into many smaller tables ( called partitions or child tables)

so, Instead of one big `osm_road_geometry`table and a complex filtering 
* there will be   6 road `child tables`  +  simple SQL ` type in ` filtering   
* OR there will be ~25  road  `child tables` and optimal solutions for the 'road_label' and 'road' layers.

for example the  ~25   road  `child tables` example :
* `road_z10` layer will be the SQL UNION of the following new `child tables`
 * road_motorway
 * road_trunk
 * road_primary
 * road_secondary
 * road_motorway_link
 * road_major_rail
 * road_major_rail_light
 * road_tertiary

PRO:
 * optimized tables for rendering - performance gain
 
CON:
 * to complex design , lot of tables  
  * need some hand made design tools,  meta tables , ....   
 * We need to add some clever post-processing and/or some advanced filtering :
  * for example the mis-tagged osm objects with:  railway=rail AND golf=hole   

RISK:
 * The current (nonofficial and not final and not matured ) advanced filtering  -> [work in progress ] ( https://github.com/ImreSamu/imposm3/commit/73ab6de78a6d5e528552d6c1d09adecba8d5b9a4 ) 


STATUS:
 *  only draft design:  [ xls ]  
(https://docs.google.com/spreadsheets/d/1ilMLJ_iXIfWJ39ucOI5VIIjwvnClem6RbY5RxZP1Ies/edit?usp=sharing)
  * 6 road child tables
  * 25 road child tables  




**MORE IMPROVED IDEA:**
 * the  ( next generation)  imposm3 will have an  Auto-partitioning  support.   
   * and will  be automatically  generating  [PG Master and Child Tables] (http://www.postgresql.org/docs/9.5/static/ddl-partitioning.html)  from the mapping definitions.
      *  by  `type`  ( = `mapping_value` )
      *  and/or  by QuadTiles

**Comment by lukasmartinelli** [comment]
> my optimization idea ( need more research , and for just the v2.x iterations )

Like always very nice proposition for the future. Mapping is already frozen now so we can support diff updates so a possible implementation of this is quite far in the future.

It's nice to see that imposm3 can be pushed in this direction!

At the moment the biggest problem is not the SQL speed with the current hardware (but I guess as OSM grows the problem will come) but the issue with the tiles that take so long (perhaps but not confirmed due to #314).

**Comment by lukasmartinelli** [comment]
Murphys law happened. The SQLite database where the tile jobs continuously get merged into by the `merge-jobs` component got corrupted.

The green ones are the tiles that I could not recover and therefore need to rerender.
Luckily I know which tiles.

<img width="669" alt="qgis_2_14_0-essen_-_missing_tiles" src="https://cloud.githubusercontent.com/assets/1288339/15267749/a14f6088-19ca-11e6-8246-fd54297dcf23.png">

**Comment by lukasmartinelli** [comment]
Rendering the lost tiles continues.

![qgis_2_14_0-essen_-_missing_tiles](https://cloud.githubusercontent.com/assets/1288339/15274595/45d31726-1ab6-11e6-91a5-0493a48b4d2a.png)

Another cool map below. This shows all z10 tiles that changed over the course of the last ten days and have been detected. Seems admin boundaries are often affected. Not sure what the reason for the big blocks of tiles that are now dirty in Russia are. Also nice to see the big mapping activity in Africa.

![changed_tiles_z10](https://cloud.githubusercontent.com/assets/1288339/15274600/773c1fa6-1ab6-11e6-8a0a-878664541eb8.png)

**Comment by sfkeller** [comment]
@lukasmartinelli wrote:

> Seems admin boundaries are often affected. Not sure what the reason for the big blocks 
> of tiles that are now dirty in Russia are. 

Weird observations: You could make a blog post about this nice "map of changes at z10". Why so many changes at boundaries? Why no "inner continental" changes in the Americas and Russia? Why so many changes in northern Russia territory?

**Comment by ImreSamu** [comment]
>Why so many changes at boundaries? 

my2c:  OSM butterfly effect :  https://en.wikipedia.org/wiki/Butterfly_effect
`The butterfly effect refers to a concept that small causes can have large effects.`

for example modifying small part of France - Schweiz border : http://www.openstreetmap.org/changeset/39252252

Has an effect of other part of the  French-administered territories ...
![france_osm_2202162](https://cloud.githubusercontent.com/assets/2217320/15275888/f1e4b8be-1ad7-11e6-9ad8-ee5c3568aec4.png)
* [Wikipedia: Overseas departments and territories of France](https://en.wikipedia.org/wiki/Overseas_departments_and_territories_of_France)
* [Map of  French Overseas Departments and Territories](https://en.wikipedia.org/wiki/Overseas_departments_and_territories_of_France#/media/File:Outre-mer_en.png)


> Seems admin boundaries are often affected

probably the 10day refresh rate for  `#admin`	 to short  

Data updates / Update frequency   based on Mapbox specification:
* V7 `#admin`	   **irregular schedule; every 2-6 months**
* V6 `#admin`	   **irregular schedule; every 2-6 months**
* V5 `#admin`	   **every few months**

**Comment by lukasmartinelli** [comment]
> for example modifying small part of France - Schweiz border : http://www.openstreetmap.org/changeset/39252252

Thank @ImreSamu for that example. Cool to see that we exactly have the same changes for admin boundaries as your example.

> Data updates / Update frequency based on Mapbox specification:
> 
> V7 #admin irregular schedule; every 2-6 months
> V6 #admin irregular schedule; every 2-6 months
> V5 #admin every few months

I guess Mapbox relates to the Natural Earth borders here not the OSM boundaries. But not sure.
Would be good to measure the impact of such a decision - e.g. how many tiles we save from rerendering.


Below are the tiles that took longer than 5 min to render. Now we are back to Friday with the lost tiles recovered. Now we do a second run with 20min rendering time upper limit and then can investigate the tiles that still won't complete.
Greenland is really not happy perhaps due to the glaciers #314 (but still only a hypothesis).
The tiles that take long on mainland are probably just data intensive tiles.

<img width="766" alt="qgis_2_14_0-essen_-_missing_tiles" src="https://cloud.githubusercontent.com/assets/1288339/15303345/95c4f950-1bb7-11e6-8715-a44d11b5172c.png">

First I will just render the tiles below and then progress to north and south pole.

<img width="818" alt="qgis_2_14_0-essen_-_missing_tiles" src="https://cloud.githubusercontent.com/assets/1288339/15303532/8f0bd376-1bb8-11e6-908e-51e69ca90387.png">

**Comment by lukasmartinelli** [comment]
Okay @manuelroth and me found one cause for the missing tiles (that is not even a problem) in the SQLite db.
This affects tiles on land where there is **no data at z8**.
For example here in Mali as you zoom in you will see a village on z10. But at z8 there is no data (0 bytes) because there is nothing to show at that zoom level. Therefore the tile pyramid is marked as missing when looking in the SQLite file (but it is not missing - there is just no data at z8 as you zoom in you will see data at z10).

This never occured before because `tl copy` does store tiles with 0 bytes in the MBTiles file while `tilelive-copy` does not.

tldr: **Empty z8 tiles on mainland are not a problem we need to solve**. The behaviour of having no tiles is completly ok. The tileserver will just serve a 404 until you hit a tile where there is data.
This doesn't affect us it just skews the tools that verify the SQLite file.

This still leaves open the tiles in the USA on Delaware where tiles are also missing.
But now we have an explanation for missing tiles on empty mainland (same is valid for southpole).


![geojson_io](https://cloud.githubusercontent.com/assets/1288339/15319666/470685b6-1c2c-11e6-9d46-0b370f6550e5.png)

**Comment by lukasmartinelli** [comment]
![geojson_io](https://cloud.githubusercontent.com/assets/1288339/15320109/154133d4-1c2f-11e6-8474-8afbe7f2d510.png)

The problem in this part of the US is that they are just very data heavy tiles due to the landuse polygons. The data is already sliced in OSM itself which is weird.

**Solution here is to give the rendering more time 30 to 40min to complete**.

![openstreetmap](https://cloud.githubusercontent.com/assets/1288339/15320118/23f03966-1c2f-11e6-9029-fa2a12f06957.png)

![openstreetmap](https://cloud.githubusercontent.com/assets/1288339/15320158/4bf917e8-1c2f-11e6-899b-bc0e35b654bf.png)

**Comment by lukasmartinelli** [comment]
Problem of these tiles here is one huge national park.
This can probably be solved with the approach mentioned by @ImreSamu.

> ST_Subdivide() http://blog.cartodb.com/subdivide-all-things/ from the blog: "Subdividing big things can make map drawing faster too, but beware: once your polygons are subdivided you’ll have turn off the polygon outlines to avoid showing the funny square boundaries in your rendered map."

However our Postgis database doesn't support the GEOS version that supports `ST_Subdivide` which is a problem because of the diffs. Need to investigate here. Subdivide looks quite comfortable since we can then just create new tables for the very large polygons.

![cursor_and_geojson_io](https://cloud.githubusercontent.com/assets/1288339/15320409/aa0f83f2-1c30-11e6-8a73-01d02f3948da.png)


![osm2vectortiles_v1_4](https://cloud.githubusercontent.com/assets/1288339/15320392/915cd878-1c30-11e6-9929-b14828c49b04.png)

**Comment by stirringhalo** [comment]
I will be doing a similar import for the planet. What sort of times are you looking at on S3 with how many cores for postgres and workers? Is this on multiple small machines (like one for Postgres and many for processing) or one large? How's the scaling to large CPU machines?

**Comment by lukasmartinelli** [comment]
> I will be doing a similar import for the planet. What sort of times are you looking at on S3 with how many cores for postgres and workers? Is this on multiple small machines (like one for Postgres and many for processing) or one large? How's the scaling to large CPU machines?

Happy to answer the question.
The more cores the better (especially for import). Rendering as well.
Make sure PostGIS has enough memory to keep the indizes in memory (32GB+ better 50GB).

Import planet with diff and index etc: 12+ hrs
Import diffs of 10 days: 8+hrs
Render planet:  From 1 minute up to 30 minutes for one z8 down to z14 subpyramid with a single worker, so at the absolute minimum 45days for single renderer. For 32 workers it takes us about a week. But it is skewed I the very few minority of the tiles (5%?) of the tiles take most of the time (60%?) of the time.

But you can scale the rendering process on the machine (4 workers and one PostGIS on 16x core machine - PostGIS is using 12 cores and each worker one).
And then you can scale to multiple hosts to bring down the time.

One large one is of course easier to operate (since you don't need to copy around the database).
If you use multiple hosts I recommend copying around the database files.

S3 is usually not the problem. But as seen above you can face some problems with data that you don't expect since on the entire world there are many possibilities for weird data :)

We hope to outline the interesting parts in the thesis as well.

**Comment by lukasmartinelli** [comment]
Created most of the extracts now. Theoretically ready to publish the new documentation (and all extracts) in #324 and finally publish vector tiles v2.0.

@klokan can you update the CDN with a new version v2.0 (we will rename mileston v1.6 into v2.0 since according to semantic versioning we break changes by changing the data format of our provided data). Size is now ~**50GB** - vacuuming the db saved 4GBs.

Links for planet.mbtiles:

- https://osm2vectortiles-downloads.os.zhdk.cloud.switch.ch/v2.0/planet.mbtiles

We will provide updates for the planet.mbtiles in a few days already again (rendering the changes from the last two weeks). But CDN is important for us to display the maps on the osm2vectortiles home page (alternatively we can also run the tileserver-gl jsut serving from the MBTiles file - will also perform well).

**Comment by stirringhalo** [comment]
Sounds great! Sorry for all the questions!

Just digging in with docker stats to better understand the architecture. I see one worker is fired up per machine (with the default docker-compose up export), and in fact you're limited by postgis performance. 

Is there a natural limit you find to scaling to more cores? Do you expect the rendering would scale to much larger machines with more cores? From your first comment, it sounds like a single worker on the same machine will saturate a postgres instance? But you mention 4 workers and one postgis on a 16 core machine?

I haven't been able to get more than one worker busy on my 4 core workstation, I ran "docker-compose scale export-worker=2; docker-compose up export-worker" but I get errors related to cannot connect to S3. I saw that mock-s3 was installed/started, rabbitmq was installed/started and two workers were started by docker, however it doesn't seem to use mock-s3. I assume there's a step I am missing!

**Comment by lukasmartinelli** [comment]
What are you planning to do? Adapting the workflow to scale the process for your own vector tiles?
Or rendering osm2vectortiles yourself?

> Is there a natural limit you find to scaling to more cores? Do you expect the rendering would scale to much larger machines with more cores? From your first comment, it sounds like a single worker on the same machine will saturate a postgres instance? But you mention 4 workers and one postgis on a 16 core machine?

Scale the export-worker until all CPUs are used and as long not too many SQL connections are open :) It does scale as you add more cores.

> I haven't been able to get more than one worker busy on my 4 core workstation, I ran "docker-compose scale export-worker=2; docker-compose up export-worker" but I get errors related to cannot connect to S3. I saw that mock-s3 was installed/started, rabbitmq was installed/started and two workers were started by docker, however it doesn't seem to use mock-s3. I assume there's a step I am missing!

Okay this means you got through the entire process of scheduling jobs as described [here](https://github.com/osm2vectortiles/osm2vectortiles/tree/master/src/generate-jobs) 😁 ?

```bash
docker-compose run generate-jobs python generate_jobs.py pyramid 5 5 8 --job-zoom=9 > jobs.json
export AMQP_URI=amqp://osm:osm@localhost:5672/ # replace localhost with docker host
cat jobs.json | pipecat publish jobs
```

Sorry we don't have many docs there. It should work with mock-s3 just checked.
You can also peek in the s3 container to check that. 

```
docker exec -it osm2vectortiles_mock-s3_1 bash
ls /data/osm2vectortiles-jobs/
```

Last route is always to pull newest master and execute `make fast` to ensure you got latest containers.

**Comment by lukasmartinelli** [comment]
BTW. about updated tiles.
Just pulled in the latest dirty tiles of the last 10 days and merged those tiles with the tiles of the previous 10 days. It turns out that the unique changed tiles of 20 days is ~ 20million tiles while unique changed tiles of 10 days has ~18 million tiles.

This confirms the hypothesis that **few tiles change very often**.

And more timing infos. Generating city extracts (using Python multiprocess pool to scale) takes only about 1 hour. Country extracts seem to take much longer.

**Comment by lukasmartinelli** [comment]
World is rendered. I am now already in rendering the 20million dirty tiles.
Perhaps we need to pull out some issues that have accumulated here and occur  later on.
